feat(storage): move Storage to use Pigeon for platform channels

cynthiajoan opened the pull request:
## Description

*Replace this paragraph with a description of what this PR is doing. If you're modifying existing behavior, describe the existing behavior, how this PR is changing it, and what motivated the change.*

## Related Issues

*Replace this paragraph with a list of issues related to this PR from the [issue database](https://github.com/flutter/flutter/issues). Indicate, which of these issues are resolved or fixed by this PR. Note that you'll have to prefix the issue numbers with flutter/flutter#.*

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`melos run analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change.
- [ ] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/firebase/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

Lyokone commented:
Hey @cynthiajoan, the code LGTM for the migration waiting for the tests to go green

russellwheatley commented on packages/firebase_storage/firebase_storage/android/src/main/java/io/flutter/plugins/firebase/storage/TaskStateChannelStreamHandler.java:
```diff
@@ -0,0 +1,114 @@
+  private final String TASK_APP_NAME = "appName";
+  private final String TASK_SNAPSHOT = "snapshot";
+
+  private static final Executor taskExecutor = Executors.newSingleThreadExecutor();
+  private Boolean destroyed = false;
+
+  private final Object pauseSyncObject = new Object();
+  private final Object resumeSyncObject = new Object();
+  private final Object cancelSyncObject = new Object();
```
These need removing, use the instances from the `FlutterFirebaseStorageTask` created. i.e pass them into the constructor for this class.

cynthiajoan replied:
done, although I feel these sync object is actually not necessary in the new way.

russellwheatley commented on packages/firebase_storage/firebase_storage/android/src/main/java/io/flutter/plugins/firebase/storage/TaskStateChannelStreamHandler.java:
```diff
@@ -0,0 +1,114 @@
+              TASK_STATE_NAME,
+              GeneratedAndroidFirebaseStorage.PigeonStorageTaskState.SUCCESS.index);
+          events.success(event);
+          //destroy();
```
`destroy()` needs to be reinstated into these [three event listeners](https://github.com/firebase/flutterfire/blob/master/packages/firebase_storage/firebase_storage/android/src/main/java/io/flutter/plugins/firebase/storage/FlutterFirebaseStorageTask.java#L268-L301) for the `StorageTask`.

cynthiajoan replied:
done

russellwheatley commented on packages/firebase_storage/firebase_storage/android/src/main/java/io/flutter/plugins/firebase/storage/TaskStateChannelStreamHandler.java:
```diff
@@ -0,0 +1,114 @@
+  }
+
+  @Override
+  public void onCancel(Object arguments) {
```
When a task is completed/cancelled/failed. This needs to be called from the `FlutterFirebaseStorageTask`, also need to clean up from the Maps holding them [here](https://github.com/firebase/flutterfire/pull/11521/files#diff-9de9c3a74a36c16b77f4d83c24fe808819acbb88a6ff4f6f59a5e882da49c438R47-R48)

russellwheatley replied:
I would put the clean up of a stream handler in the destroy method here: https://github.com/firebase/flutterfire/pull/11521/files#diff-b6e5a6bea234c57c8a9a1bf39f759509722c844490626f89f7f7575f2ecf8169R140

cynthiajoan replied:
done

russellwheatley replied:
The `onCancel()` method needs to be called whenever the task is completed/cancelled/failed. We also need to remove it from the `eventChannels` map and `streamHandlers`.

Making a function like below:
```java
private void removeEventListener(String identifier) {
      eventChannels.get(identifier).setStreamHandler(null);
      streamHandlers.get(identifier).onCancel(null);
  }
```
and call it from `destroy()` method in the `FlutterFirebaseStorageTask` class. Otherwise, the only time the stream handlers are removed are when the Flutter engine is detached.

You can also remove [this `destroy()`](https://github.com/firebase/flutterfire/blob/870432c6336838209c63965ba2107699343e95c0/packages/firebase_storage/firebase_storage/android/src/main/java/io/flutter/plugins/firebase/storage/TaskStateChannelStreamHandler.java#L94) method because the task is already destroyed in `onDetachfromEngine`, `didReinitializeFirebaseCore` & when we call `destroy()` in the `addOnFailureListener`, `addOnCanceledListener` & `addOnSuccessListener` listeners which are the times it needs to be destroyed. The issue might be that we call `flutterTask.destroy()` on a task that has already been removed from the `inProgressTasks` static array.

In fact, it might be worth adding this line to the top of the `destroy()` method:

```java
 void destroy() {
// The line below:
    if(destroyed) return;
    destroyed = true;

    synchronized (inProgressTasks) {
      if (storageTask.isInProgress() || storageTask.isPaused()) {
        storageTask.cancel();
      }
      inProgressTasks.remove(handle);
    }

    synchronized (cancelSyncObject) {
      cancelSyncObject.notifyAll();
    }

    synchronized (pauseSyncObject) {
      pauseSyncObject.notifyAll();
    }

    synchronized (resumeSyncObject) {
      resumeSyncObject.notifyAll();
    }
  }

```

russellwheatley commented on packages/firebase_storage/firebase_storage/android/src/main/java/io/flutter/plugins/firebase/storage/TaskStateChannelStreamHandler.java:
```diff
@@ -0,0 +1,114 @@
+  @Override
+  public void onListen(Object arguments, EventSink events) {
+    androidTask.addOnProgressListener(
+        taskSnapshot -> {
```
Pass down `taskExecutor` from the `FlutterFirebaseStorageTask` and put as first argument in `addOnProgressListener`, `addOnPausedListener`, `addOnSuccessListener`, `addOnCanceledListener` & `addOnFailureListener` . As it was previously: https://github.com/firebase/flutterfire/blob/master/packages/firebase_storage/firebase_storage/android/src/main/java/io/flutter/plugins/firebase/storage/FlutterFirebaseStorageTask.java#L244

cynthiajoan replied:
When using taskExecutor to run it will throw exception of java.lang.RuntimeException: Methods marked with @UiThread must be executed on the main thread.

russellwheatley commented on packages/firebase_storage/firebase_storage/android/src/main/java/io/flutter/plugins/firebase/storage/FlutterFirebaseStoragePlugin.java:
```diff
@@ -133,14 +135,61 @@ public void onAttachedToEngine(@NonNull FlutterPluginBinding binding) {
+    GeneratedAndroidFirebaseStorage.FirebaseStorageHostApi.setup(messenger, null);
     channel = null;
+    messenger = null;
+    removeEventListeners();
```
`removeEventListeners()` also needs to be called within `didReinitializeFirebaseCore()` method.

cynthiajoan replied:
done

russellwheatley reviewed (changes requested):
Thank you for the PR, @cynthiajoan! I hope you don't mind, but I found some issue that ought to be addressed for updating. Also - in the `FlutterFirebaseStorageTask.java`, `FlutterFirebaseStoragePlugin.java` & `TaskStateChannelStreamHandler.java`, there are a number of warnings that ought to be addressed. For example, methods that are not used can be removed if no longer needed.

I'm going to look at the Apple side of things next, and see what might need addressing there 😄

russellwheatley commented on packages/firebase_storage/firebase_storage/ios/Classes/FLTFirebaseStoragePlugin.m:
```diff
@@ -61,39 +61,47 @@ typedef NS_ENUM(NSUInteger, FLTFirebaseStorageStringType) {
 @interface FLTFirebaseStoragePlugin ()
-@property(nonatomic, retain) FlutterMethodChannel *channel;
+@property(nonatomic, retain) FlutterMethodChannel *storage_method_channel;
+//@property(strong, nonatomic) PigeonStorageFirebaseApp *appleArguments;
```
can remove if not needed

cynthiajoan replied:
done

russellwheatley commented on packages/firebase_storage/firebase_storage/ios/Classes/FLTFirebaseStoragePlugin.m:
```diff
@@ -134,91 +145,174 @@ - (void)cleanupWithCompletion:(void (^)(void))completion {
+//                                              andOptionalNSError:error]);
+//   };
+
+//   FLTFirebaseMethodCallResult *methodCallResult =
+//       [FLTFirebaseMethodCallResult createWithSuccess:flutterResult andErrorBlock:errorBlock];
+
+//   if ([@"Storage#useEmulator" isEqualToString:call.method]) {
+//     [self useEmulator:call.arguments withMethodCallResult:methodCallResult];
+//   } else if ([@"Reference#delete" isEqualToString:call.method]) {
+//     [self referenceDelete:call.arguments withMethodCallResult:methodCallResult];
+//   } else if ([@"Reference#getDownloadURL" isEqualToString:call.method]) {
+//     [self referenceGetDownloadUrl:call.arguments withMethodCallResult:methodCallResult];
+//   } else if ([@"Reference#getMetadata" isEqualToString:call.method]) {
+//     [self referenceGetMetadata:call.arguments withMethodCallResult:methodCallResult];
+//   } else if ([@"Reference#getData" isEqualToString:call.method]) {
+//     [self referenceGetData:call.arguments withMethodCallResult:methodCallResult];
+//   } else if ([@"Reference#list" isEqualToString:call.method]) {
+//     [self referenceList:call.arguments withMethodCallResult:methodCallResult];
+//   } else if ([@"Reference#listAll" isEqualToString:call.method]) {
+//     [self referenceListAll:call.arguments withMethodCallResult:methodCallResult];
+//   } else if ([@"Reference#updateMetadata" isEqualToString:call.method]) {
+//     [self referenceUpdateMetadata:call.arguments withMethodCallResult:methodCallResult];
+//   } else if ([@"Task#startPutData" isEqualToString:call.method]) {
+//     [self taskStartPutData:call.arguments withMethodCallResult:methodCallResult];
+//   } else if ([@"Task#startPutString" isEqualToString:call.method]) {
+//     [self taskStartPutString:call.arguments withMethodCallResult:methodCallResult];
+//   } else if ([@"Task#startPutFile" isEqualToString:call.method]) {
+//     [self taskStartPutFile:call.arguments withMethodCallResult:methodCallResult];
+//   } else if ([@"Task#pause" isEqualToString:call.method]) {
+//     [self taskPause:call.arguments withMethodCallResult:methodCallResult];
+//   } else if ([@"Task#resume" isEqualToString:call.method]) {
+//     [self taskResume:call.arguments withMethodCallResult:methodCallResult];
+//   } else if ([@"Task#cancel" isEqualToString:call.method]) {
+//     [self taskCancel:call.arguments withMethodCallResult:methodCallResult];
+//   } else if ([@"Task#writeToFile" isEqualToString:call.method]) {
+//     [self taskWriteToFile:call.arguments withMethodCallResult:methodCallResult];
+//   } else {
+//     flutterResult(FlutterMethodNotImplemented);
+//   }
+// }
```
can remove

cynthiajoan replied:
done

russellwheatley commented on packages/firebase_storage/firebase_storage/ios/Classes/FLTFirebaseStoragePlugin.m:
```diff
@@ -227,136 +321,256 @@ - (void)referenceGetDownloadUrl:(id)arguments
+  NSLog(@"taskPauseApp");
 }
 
-- (void)taskStartPutFile:(id)arguments withMethodCallResult:(FLTFirebaseMethodCallResult *)result {
-  [self startFIRStorageObservableTaskForArguments:arguments
-                    andFLTFirebaseStorageTaskType:FLTFirebaseStorageTaskTypeFile];
-  result.success(nil);
+- (void)taskResumeApp:(PigeonStorageFirebaseApp *)app
+               handle:(NSNumber *)handle
+           completion:(void (^)(NSDictionary<NSString *, id> *_Nullable,
+                                FlutterError *_Nullable))completion {
+  // TODO
+  NSLog(@"taskResumeApp");
 }
 
-- (void)taskWriteToFile:(id)arguments withMethodCallResult:(FLTFirebaseMethodCallResult *)result {
-  [self startFIRStorageObservableTaskForArguments:arguments
-                    andFLTFirebaseStorageTaskType:FLTFirebaseStorageTaskTypeDownload];
-  result.success(nil);
+- (void)taskCancelApp:(PigeonStorageFirebaseApp *)app
+               handle:(NSNumber *)handle
+           completion:(void (^)(NSDictionary<NSString *, id> *_Nullable,
+                                FlutterError *_Nullable))completion {
+  // TODO
+  NSLog(@"taskCancelApp");
 }
```
These need implementing. These are the methods called by Pigeon.

cynthiajoan replied:
done

russellwheatley commented on packages/firebase_storage/firebase_storage/ios/Classes/FLTFirebaseStoragePlugin.m:
```diff
@@ -227,136 +321,256 @@ - (void)referenceGetDownloadUrl:(id)arguments
+  NSLog(@"taskCancelApp");
 }
 
 - (void)taskPause:(id)arguments withMethodCallResult:(FLTFirebaseMethodCallResult *)result {
```
`taskPause`, `taskCancel` & `taskResume` are the previous implementation and should be replaced with the above noted methods used by generated Pigeon code.

cynthiajoan replied:
done

russellwheatley commented on packages/firebase_storage/firebase_storage/ios/Classes/FLTTaskStateChannelStreamHandler.m:
```diff
@@ -0,0 +1,78 @@
+  return nil;
+}
+
+- (FlutterError *)onCancelWithArguments:(id)arguments {
```
This needs to be called and cleaned up when task has been resolved as mentioned for android.

cynthiajoan replied:
done

russellwheatley commented on packages/firebase_storage/firebase_storage/ios/Classes/FLTFirebaseStoragePlugin.m:
```diff
@@ -61,39 +61,47 @@ typedef NS_ENUM(NSUInteger, FLTFirebaseStorageStringType) {
   dispatch_queue_t _callbackQueue;
   bool hasEmulatorBooted;
+  NSObject<FlutterBinaryMessenger> *_binaryMessenger;
+  NSMutableDictionary<NSString *, FlutterEventChannel *> *_eventChannels;
+  NSMutableDictionary<NSString *, NSObject<FlutterStreamHandler> *> *_streamHandlers;
```
These need to be cleaned up in `cleanupWithCompletion` method.

cynthiajoan replied:
done

russellwheatley commented:
I think we should bring back these tests for this PR: https://github.com/firebase/flutterfire/blob/master/tests/integration_test/firebase_storage/task_e2e.dart

We skip because of how flakey they are on the emulator but worth checking for this as the internal implementation has now changed for tasks that pause/resume/cancel.

cynthiajoan commented:
> I think we should bring back these tests for this PR: https://github.com/firebase/flutterfire/blob/master/tests/integration_test/firebase_storage/task_e2e.dart
> 
> We skip because of how flakey they are on the emulator but worth checking for this as the internal implementation has now changed for tasks that pause/resume/cancel.

I brought part of the tests back, it's still flaky for some of them, I'm letting them running to see what the results would be.

russellwheatley reviewed (changes requested):
I've [made one comment here](https://github.com/firebase/flutterfire/pull/11521#discussion_r1366945332) but other than that, LGTM

cynthiajoan commented:
> I've [made one comment here](https://github.com/firebase/flutterfire/pull/11521#discussion_r1366945332) but other than that, LGTM

Thanks @russellwheatley, addressed the comment.
